test/apps/iodemo/run_io_demo.sh

```diff
@@ -414,238 +417,238 @@ make_scripts()
-			cmd_prefix="stdbuf -e0 -o0 timeout -s 9 $((duration + client_wait_time))s  env IODEMO_ROLE=server_${i} "
 			echo ${log_file}
 			cat >>${command_file} <<-EOF
-			function start_server_${i}() {
-				${cmd_prefix} \\
-				    ${iodemo_exe} \\
-				        ${iodemo_server_args} -p ${port_num} \\
-				        ${log_redirect} ${log_file} &
-			}
+				function start_server_${i}() {
+				    mkdir -p ${log_dir}
+				    env IODEMO_ROLE=server_${i} ${cmd_prefix} \\
+				        ${iodemo_exe} \\
+				            ${iodemo_server_args} -p ${port_num} \\
+				            ${log_redirect} ${log_file} &
+				}
+
 				EOF
 		done
 
-
-		# Add client commands
-		cat >>${command_file} <<-EOF
-
-			# Client commands
-			EOF
-
+		# Add client start functions
+		cmd_prefix="stdbuf -e0 -o0 timeout -s 9 ${duration}s"
 		for ((i=0;i<num_clients_per_host[${host}];++i))
 		do
 			log_file=${log_dir}/$(printf "iodemo_%s_client_%02d.log" ${host} $i)
-			cmd_prefix="stdbuf -e0 -o0 timeout -s 9 ${duration}s env IODEMO_ROLE=client_${i} "
 			echo ${log_file}
 			cat >>${command_file} <<-EOF
-			function start_client_${i}() {
-				${cmd_prefix} \\
-				    ${iodemo_exe} \\
-				        ${iodemo_client_args} ${client_connect_list} \\
-				        ${log_redirect} ${log_file} &
-			}
+				function start_client_${i}() {
+				    mkdir -p ${log_dir}
+				    env IODEMO_ROLE=client_${i} ${cmd_prefix} \\
+				        ${iodemo_exe} \\
+				            ${iodemo_client_args} ${client_connect_list} \\
+				            ${log_redirect} ${log_file} &
+				}
+
 				EOF
 		done
 
+		# 'run_all' will start all servers, then clients, then wait for finish
 		cat >>${command_file} <<-EOF
-		main() {
-		EOF
+			run_all() {
+			    ${set_verbose}
+
+			    # kill existing processes and trap signals
+			    kill_iodemo
+			    trap signal_handler INT TERM
+
+			    set_env_vars
+
+			    echo "Starting servers"
+				EOF
+
 		for ((i=0;i<${num_servers_per_host[${host}]};++i))
 		do
-			cat >>${command_file} <<-EOF
-			start_server_${i}
-				EOF
+			echo "    start_server_${i}" >>${command_file}
 		done
 
-		# Add short sleep to let servers start running
 		cat >>${command_file} <<-EOF
-			# Wait for servers to start
-			sleep ${client_wait_time}
-		EOF
 
-		for ((i=0;i<num_clients_per_host[${host}];++i))
-		do
-			cat >>${command_file} <<-EOF
-			start_client_${i}
+			    # Wait for servers to start
+			    sleep ${client_wait_time}
+
+			    echo "Starting clients"
 				EOF
+
+		for ((i=0;i<${num_clients_per_host[${host}]};++i))
+		do
+			echo "    start_client_${i}" >>${command_file}
 		done
 
-		# Add wait commands
 		cat >>${command_file} <<-EOF
 
-			# Wait for background processes
-			wait ${wait_redirect}
-			echo "Test finished"
-		}
+			    # Wait for background processes
+			    wait ${wait_redirect}
+			    echo "Test finished"
+			}
+
 			EOF
-		cat >>${command_file} <<-'EOF'
-		parse_args "$@"
-		kill_iodemo
-		main
+
+		cat >>${command_file} <<-EOF
```

Comment:
you can use ```<<-'EOF'``` and no need to wrap ```\$```

Reply:
in some place i want to replace vars from external script